### ISSUE TITLE
Execution time for params.get increases exponentially based on context mount order

### ISSUE THREAD

**Issue by enragedginger** [scalatra/scalatra#658]
I'm using Scala 2.11.8 and Scalatra 2.4.1. Because I do not have unique base paths for each "servlet," I am using `ScalatraFilter` instances and mounting all of them to `"/*`. Essentially, when a request comes in, if I call `params.get("userId")` from within my servlet instance (after a route has been matched), and `userId` is on the request, then the value is found almost immediately. However, if `userId` is not present on the request, the amount of time required to return `None` increases exponentially with respect to the number of `ScalatraFilter` instances in the stack above the current instance.

The execution times looked something like this:
- 8 servlets / filters ~8 millis
- 13 servlets / filters ~76 millis
- 16 servlets / filters ~700 millis
- 18 servlets / filters ~2.3 seconds
- 20 servlets / filters ~3.8 seconds
- 23 servlets / filters ~30 seconds
- 27 servlets / filters ~4.5 minutes

Essentially, the order in which one mounts `ScalatraFilter` instances to the context via `context.mount` appears to effect the execution time of seemingly unrelated operations in a surprising way. Is there a fix for this?

**Comment by takezoe** [comment]
I expect that `params.get("userId")` is not related. Did you measure it?

According to the number of mounted filter, routing becomes slow progressively. In particular, very slow if it can't find matched routing because it have to explore all routing. I've wanted to fix this problem long time, but I don't have a good idea for now.

**Comment by enragedginger** [comment]
Surprisingly, yes, `params.get("userId")` was the source of the slowness, and not the route matching. The times listed above are the measurements of execution times of that specific statement and not the entire endpoint runtime.

I tried about 15 different things within my app and couldn't resolve this specific issue. Somewhere along the line, I did an `sbt clean compile` which seemed to make the problem go away and I could no longer reproduce it. I'll reopen this issue if I can find a way to deterministically reproduce the issue.

Thank you!
